### PR TITLE
PI-83 updates to handle release processing

### PR DESCRIPTION
**Pull request by stevomcallister**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDate;
+
+@Data
+@Builder
+@NoArgsConstructor
```
Comment (trozzomoj) [inline src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java:12]: I m not sure you need this. Doe the NewRelase bein construct even with an empty constructor ? Or Actualrelease , releasetype and institution are mandatory fields ? If is the case. Remove annotation , make the fields final.

File: src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    private Long id;
+    private LocalDate actualReleaseDate;
+
+    public NewRelease(LocalDate actualReleaseDate, String releaseType, String institution) {
+        this.actualReleaseDate = actualReleaseDate;
+        this.releaseType = releaseType;
+        this.institution = institution;
+    }
+
+    private String releaseType;
```
Comment (trozzomoj) [inline src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java:24]: I think filed declaration would look better up before the constructor.

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java [hunk 3]
```diff
@@ -48,6 +53,7 @@
 @Service
 @Slf4j
 public class CustodyService {
+    public static final String CONTEXT = "prison-to-probation";
```
Comment (trozzomoj) [inline src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:56]: I this should be placed elsewhere if is needed but injected ${strings.key} .

File: src/main/java/uk/gov/justice/digital/delius/data/api/OffenderReleasedNotification.java [hunk 4]
```diff
@@ -18,6 +18,10 @@
     @ApiModelProperty(value = "The Prison institution code in NOMIS the offender was released from", required = true, example = "MDI")
     private String nomsPrisonInstitutionCode;
 
+    @NotBlank(message = "Missing a NOMS release reason code")
```
Comment (trozzomoj) [inline src/main/java/uk/gov/justice/digital/delius/data/api/OffenderReleasedNotification.java:21]: Not sure but seems these validations are not tested ?

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java [hunk 5]
```diff
@@ -460,4 +485,11 @@ static PrisonLocationUpdateSuccess noUpdateRequired(final List<Custody> custodyR
         all.putAll(map2);
         return Map.copyOf(all);
     }
+
+    IntegrationContext getContext(String contextName) {
+        var context = deliusIntegrationContextConfig.getIntegrationContexts().get(contextName);
+        return ofNullable(context).orElseThrow(
+            () -> new BadRequestException("IntegrationContext does not exist for: " + contextName)
```
Comment (trozzomoj) [inline src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:492]: This validation is not tested (?)

**Review by trozzomoj — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java (hunk 5 above)
Comment (marcus-bcl) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:492]: Also should this be an `IllegalStateException`, or something similar?

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java [hunk 6]
```diff
@@ -273,6 +287,17 @@ public Custody offenderReleased(final String nomsNumber, final OffenderReleasedN
...
     }
 
+    private void addRelease(OffenderReleasedNotification releasedNotification, Long eventId, String crn) {
+        var context = getContext(CONTEXT);
+        var releaseMapping = context.getReleaseTypeMapping();
+        var deliusReleaseType = ofNullable(releaseMapping.getReasonToReleaseType().get(releasedNotification.getReason())).orElseThrow(
+            () -> new BadRequestException("Release Type mapping from reason does not exist for: " + releasedNotification.getReason())
+        );
+
+        val newRelease = new NewRelease(releasedNotification.getReleaseDate(), deliusReleaseType, releasedNotification.getNomsPrisonInstitutionCode());
```
Comment (marcus-bcl) [inline src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:297]: Just an option, but you could make use of the `@Builder` here and ditch the constructor if you wanted.
```suggestion
        val newRelease = NewRelease.builder()
                .releaseType(deliusReleaseType)
                .actualReleaseDate(releasedNotification.getReleaseDate())
                .institution(releasedNotification.getNomsPrisonInstitutionCode())
                .build();
```

**Comment by marcus-bcl** [comment]
Looks good to me, happy to approve once @trozzomoj 's happy.  Did we manage to do an end-to-end test?

**Comment by stevomcallister** [comment]
>

File: src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java (hunk 2 above)
Comment (stevomcallister) [reply src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java:24]: oops - I blame intelliJ

File: src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java (hunk 1 above)
Comment (stevomcallister) [reply src/main/java/uk/gov/justice/digital/delius/data/api/deliusapi/NewRelease.java:12]: yup sorted that.

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java (hunk 3 above)
Comment (trozzomoj) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:56]: actually seems is used just locally (?) i case just change the visibility to private.

Comment (stevomcallister) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java:56]: ok

**Review by trozzomoj — APPROVED** [review]

**Review by marcus-bcl — DISMISSED** [review]

**Review by marcus-bcl — APPROVED** [review]

**Pull request merged**